src/jupyterhub_config/custom_spawner.py

```diff
@@ -126,4 +133,38 @@ def _ensure_user_jupyter_directory(self, user_dir: Path):
 
         self.environment['JUPYTER_CONFIG_DIR'] = str(jupyter_dir)
         self.environment['JUPYTER_RUNTIME_DIR'] = str(jupyter_runtime_dir)
-        self.environment['JUPYTER_DATA_DIR'] = str(juputer_data_dir)
\ No newline at end of file
+        self.environment['JUPYTER_DATA_DIR'] = str(juputer_data_dir)
+
+    def _ensure_virtual_environment(self, user_env_dir: Path):
+        """
+        Ensure the user's virtual environment exists. If it does not exist, it is
+        created with the system site-packages included.
+        """
+        if not user_env_dir.exists():
+            user_env_dir.mkdir(parents=True)
+            self.log.info(f'Creating virtual environment for {self.user.name}')
+            try:
+                # Create a virtual environment with system site-packages access
+                subprocess.run(['python3', '-m', 'venv', str(user_env_dir), '--system-site-packages'],
+                               check=True)
```

Comment:
https://docs.python.org/3/library/venv.html#venv.EnvBuilder.create might be better here

Reply:
👍

Reply:
:+1: